File: C_plus_plus/Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        queue<pair<TreeNode *, int>> que;
        vector<vector<int> > result;
        vector<int> tmp;
        int level = 0;
        que.push(make_pair(root,0));
        while(!que.empty()){
            pair<TreeNode *, int> p = que.front();
            que.pop();
            if(p.first){
                if(p.second>level){
                    result.insert(result.begin(),tmp);
                    tmp.clear();
                    level = p.second;
                }
                tmp.push_back(p.first->val);
                que.push(make_pair(p.first->left,p.second+1));
                que.push(make_pair(p.first->right,p.second+1));
            }
        }
        if(tmp.size()) result.insert(result.begin(),tmp);
        return result;        
    }
};

File: java/Binary_Tree_Level_Order_Traversal_II.java
// Concise solution !
// Make each level to a list, and put it to the result list's head.
// Time is O(n), space is O(n).

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new LinkedList<List<Integer>>();
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        que.offer(root);
        while (que.peek() != null) {
            List<Integer> list = new ArrayList<Integer>();
            for (int n = que.size(); n > 0; n--) {
                TreeNode node = que.poll();
                list.add(node.val);
                if (node.left != null) que.offer(node.left);
                if (node.right != null) que.offer(node.right);
            }
            result.add(0,list);
        }
        return result;
    }
}
